types/next/app.d.ts

```diff
@@ -1,19 +1,46 @@
 import * as React from "react";
-import { NextContext } from ".";
+import { NextContext, NextComponentType } from ".";
 import { RouterProps, DefaultQuery } from "./router";
 
-export interface AppComponentProps<Q = DefaultQuery> {
-    Component: React.ComponentType<any>;
+// Deprecated
+export type AppComponentProps = AppProps;
+export type AppComponentContext = NextAppContext;
+// End Deprecated
+
+/**
+ * Context passed to App.getInitialProps.
+ * Component is dynamic - cannot infer type.
+ *
+ * @template Q Query object schema.
+ */
+export interface NextAppContext<Q = DefaultQuery> {
```

Comment:
Reordered interfaces for consistency aligned with the order in `Document`. The only difference is the `Component: NextComponentType<any, any, NextContext<Q>>` type in both.